File: python/sol/farfield.py
# -*- coding: utf-8 -*-
"""
farfield.py
"""

import math, cmath
import numpy as np
from numba import jit, prange
import sol.Fnode

# 遠方界計算の準備
# PBCのときは面積=0として飛ばす
# 並列化は時間が増えるので不可
#@jit(cache=True, nogil=True, parallel=True, nopython=True)
@jit(cache=True, nopython=True)
def setup(
    Nx, Ny, Nz, Xn, Yn, Zn, Xc, Yc, Zc, Freq2, pbc,
    cEx, cEy, cEz, cHx, cHy, cHz,
    Ni, Nj, Nk, N0, NN):

    nfreq = len(Freq2)
    num = 2 * ((Nx * Ny) + (Ny * Nz) + (Nz * Nx))

    fSurface = np.zeros((7, num), 'f8')
    cSurface = np.zeros((6, num, nfreq), 'c8')

    # 作業配列
    cex = np.zeros((2, 2), 'c8')
    cey = np.zeros((2, 2), 'c8')
    cez = np.zeros((2, 2), 'c8')
    chx = np.zeros((2, 2), 'c8')
    chy = np.zeros((2, 2), 'c8')
    chz = np.zeros((2, 2), 'c8')

    # counter
    n = 0

    # X surface
    for side in range(2):
        i = [0, Nx][side]
        for j in range(Ny):
            for k in range(Nz):
                for ifreq in range(len(Freq2)):
                    for jm in range(2):
                        for km in range(2):
                            _, cey[jm][km], cez[jm][km] = \
                                sol.Fnode.e(ifreq, i, j + jm, k + km, \
                                    Nx, Ny, Nz, cEx, cEy, cEz, Ni, Nj, Nk, N0, NN)#NEx, NEy, NEz)
                            _, chy[jm][km], chz[jm][km] = \
                                sol.Fnode.h(ifreq, i, j + jm, k + km, \
                                    cHx, cHy, cHz, Ni, Nj, Nk, N0, NN)#NHx, NHy, NHz)
                    cSurface[0, n, ifreq] = complex(0, 0)
                    cSurface[1, n, ifreq] = np.sum(cey) / 4
                    cSurface[2, n, ifreq] = np.sum(cez) / 4
                    cSurface[3, n, ifreq] = complex(0, 0)
                    cSurface[4, n, ifreq] = np.sum(chy) / 4
                    cSurface[5, n, ifreq] = np.sum(chz) / 4
                # 周波数によらない形状データ
                fSurface[0, n] = [-1, +1][side]
                fSurface[1, n] = 0
                fSurface[2, n] = 0
                fSurface[3, n] = Xn[i]
                fSurface[4, n] = Yc[j]
                fSurface[5, n] = Zc[k]
                fSurface[6, n] = (Yn[j + 1] - Yn[j]) * (Zn[k + 1] - Zn[k]) * (1 - pbc[0])
                n += 1

    # Y surface
    for side in range(2):
        j = [0, Ny][side]
        for k in range(Nz):
            for i in range(Nx):
                for ifreq in range(len(Freq2)):
                    for km in range(2):
                        for im in range(2):
                            cex[km][im], _, cez[km][im] = \
                                sol.Fnode.e(ifreq, i + im, j, k + km, \
                                    Nx, Ny, Nz, cEx, cEy, cEz, Ni, Nj, Nk, N0, NN)#NEx, NEy, NEz)
                            chx[km][im], _, chz[km][im] = \
                                sol.Fnode.h(ifreq, i + im, j, k + km, \
                                    cHx, cHy, cHz, Ni, Nj, Nk, N0, NN)#NHx, NHy, NHz)
                    cSurface[0, n, ifreq] = np.sum(cex) / 4
                    cSurface[1, n, ifreq] = complex(0, 0)
                    cSurface[2, n, ifreq] = np.sum(cez) / 4
                    cSurface[3, n, ifreq] = np.sum(chx) / 4
                    cSurface[4, n, ifreq] = complex(0, 0)
                    cSurface[5, n, ifreq] = np.sum(chz) / 4
                # 周波数によらない形状データ
                fSurface[0, n] = 0
                fSurface[1, n] = [-1, +1][side]
                fSurface[2, n] = 0
                fSurface[3, n] = Xc[i]
                fSurface[4, n] = Yn[j]
                fSurface[5, n] = Zc[k]
                fSurface[6, n] = (Zn[k + 1] - Zn[k]) * (Xn[i + 1] - Xn[i]) * (1 - pbc[1])
                n += 1

    # Z surface
    for side in range(2):
        k = [0, Nz][side]
        for i in range(Nx):
            for j in range(Ny):
                for ifreq in range(len(Freq2)):
                    for im in range(2):
                        for jm in range(2):
                            cex[im][jm], cey[im][jm], _ = \
                                sol.Fnode.e(ifreq, i + im, j + jm, k, \
                                    Nx, Ny, Nz, cEx, cEy, cEz, Ni, Nj, Nk, N0, NN)#NEx, NEy, NEz)
                            chx[im][jm], chy[im][jm], _ = \
                                sol.Fnode.h(ifreq, i + im, j + jm, k, \
                                    cHx, cHy, cHz, Ni, Nj, Nk, N0, NN)#NHx, NHy, NHz)
                    cSurface[0, n, ifreq] = np.sum(cex) / 4
                    cSurface[1, n, ifreq] = np.sum(cey) / 4
                    cSurface[2, n, ifreq] = complex(0, 0)
                    cSurface[3, n, ifreq] = np.sum(chx) / 4
                    cSurface[4, n, ifreq] = np.sum(chy) / 4
                    cSurface[5, n, ifreq] = complex(0, 0)
                # 周波数によらない形状データ
                fSurface[0, n] = 0
                fSurface[1, n] = 0
                fSurface[2, n] = [-1, +1][side]
                fSurface[3, n] = Xc[i]
                fSurface[4, n] = Yc[j]
                fSurface[5, n] = Zn[k]
                fSurface[6, n] = (Xn[i + 1] - Xn[i]) * (Yn[j + 1] - Yn[j]) * (1 - pbc[2])
                n += 1

    assert(n == num)
    #print(np.sum(fSurface))
    #print(np.sum(cSurface))

    return fSurface, cSurface

@jit(cache=True, nogil=True, parallel=True, nopython=True)
#@jit(cache=True, nopython=True)
def calc(ifreq, theta, phi, ffctr, k, fSurface, cSurface):

    # sin, cos
    cost = math.cos(math.radians(theta))
    sint = math.sin(math.radians(theta))
    cosp = math.cos(math.radians(phi))
    sinp = math.sin(math.radians(phi))

    # (r1, t1, p1) : (r, theta, phi)方向の単位ベクトル
    r1 = np.array([+ sint * cosp, + sint * sinp, + cost])
    t1 = np.array([+ cost * cosp, + cost * sinp, - sint])
    p1 = np.array([- sinp,        + cosp,          0   ])

    # ポテンシャルL/N初期化(複素数ベクトル)
    pl = np.zeros(3, 'c8')
    pn = np.zeros(3, 'c8')

    # 境界要素に関するループ
    for n in prange(fSurface.shape[1]):
        # 境界要素
        nv = fSurface[0:3, n]
        pv = fSurface[3:6, n]
        ds = fSurface[6,   n]

        # 電磁界ベクトル
        ev = cSurface[0:3, n, ifreq]
        hv = cSurface[3:6, n, ifreq]

        # Z0 * J = n X (Z0 * H)
        cj = np.cross(nv, hv)

        # M = -n X E
        cm = - np.cross(nv, ev)

        # exp(jkr * r) * dS
        #expds = ds * cmath.exp(complex(0, k * np.dot(r1, pv)))
        rr = (r1[0] * pv[0]) + (r1[1] * pv[1]) + (r1[2] * pv[2])
        expds = ds * cmath.exp(complex(0, k * rr))

        # L += M * exp(jkr * r) * dS
        pl += cm * expds

        # Z0 * N += (Z0 * J) * exp(jkr * r) * dS
        pn += cj * expds

    # Z0 * N-theta, Z0 * N-phi (np.dotはjit不可)
    #pnt = np.dot(t1, pn)
    #pnp = np.dot(p1, pn)
    pnt = (t1[0] * pn[0]) + (t1[1] * pn[1]) + (t1[2] * pn[2])
    pnp = (p1[0] * pn[0]) + (p1[1] * pn[1]) + (p1[2] * pn[2])

    # L-theta, L-phi
    #plt = np.dot(t1, pl)
    #plp = np.dot(p1, pl)
    plt = (t1[0] * pl[0]) + (t1[1] * pl[1]) + (t1[2] * pl[2])
    plp = (p1[0] * pl[0]) + (p1[1] * pl[1]) + (p1[2] * pl[2])

    # F-theta, F-phi
    etheta = ffctr * (pnt + plp)
    ephi   = ffctr * (pnp - plt)

    return etheta, ephi

# 遠方界因子
def factor(ifreq, k, Pin, nfeed, matchingloss, ETA0):

    ffctr = 0
    if nfeed > 0:
        # 給電点 (post only)
        s = 0
        for ifeed in range(nfeed):
            s += 0.5 * Pin[ifreq, ifeed, matchingloss]
        ffctr = k / math.sqrt(8 * math.pi * ETA0 * s)
    else:
        # 平面波入射 (solver + post)
        einc = 1
        ffctr = k / (einc * math.sqrt(4 * math.pi))

    return ffctr

# 遠方界成分
# etheta, ephi : 複素数
def farComponent(etheta, ephi):

    e = np.zeros(7, 'f8')
    
    # abs
    e[0] = math.sqrt(abs(etheta)**2 + abs(ephi)**2)

    # theta/phi
    e[1] = abs(etheta)
    e[2] = abs(ephi)

    # major/minor
    tmp1 = abs(etheta**2) + abs(ephi**2)
    tmp2 = abs(etheta**2 + ephi**2)
    e[3] = math.sqrt((tmp1 + tmp2) / 2)
    e[4] = math.sqrt(max(tmp1 - tmp2, 0) / 2)

    # RHCP/LHCP
    e[5] = abs(etheta + 1j * ephi) / math.sqrt(2)
    e[6] = abs(etheta - 1j * ephi) / math.sqrt(2)

    return e

File: sol_sycl_dm/farfield.c
/*
farfield.c

far field
*/

#include "ofd.h"
#include "complex.h"
#include "ofd_prototype.h"

void farfield(int ifreq, double theta, double phi, double ffctr, d_complex_t *etheta, d_complex_t *ephi)
{
	// wave number

	const double kwave = (2 * PI * Freq2[ifreq]) / C;

	// unit vector in r, theta, phi

	double r1[3], t1[3], p1[3];

	const double cos_t = cos(theta * DTOR);
	const double sin_t = sin(theta * DTOR);
	const double cos_p = cos(phi   * DTOR);
	const double sin_p = sin(phi   * DTOR);

	r1[0] = +sin_t * cos_p;
	r1[1] = +sin_t * sin_p;
	r1[2] = +cos_t;

	t1[0] = +cos_t * cos_p;
	t1[1] = +cos_t * sin_p;
	t1[2] = -sin_t;

	p1[0] = -sin_p;
	p1[1] = +cos_p;
	p1[2] = 0;

	d_complex_t plx = d_complex(0, 0);
	d_complex_t ply = d_complex(0, 0);
	d_complex_t plz = d_complex(0, 0);
	d_complex_t pnx = d_complex(0, 0);
	d_complex_t pny = d_complex(0, 0);
	d_complex_t pnz = d_complex(0, 0);

	d_complex_t *ex = SurfaceEx[ifreq];
	d_complex_t *ey = SurfaceEy[ifreq];
	d_complex_t *ez = SurfaceEz[ifreq];
	d_complex_t *hx = SurfaceHx[ifreq];
	d_complex_t *hy = SurfaceHy[ifreq];
	d_complex_t *hz = SurfaceHz[ifreq];

	for (int n = 0; n < NSurface; n++) {
		surface_t *p = &Surface[n];

		// Z0 * J = n X (Z0 * H)
		const d_complex_t cjx = d_sub(d_rmul(+p->ny, hz[n]), d_rmul(+p->nz, hy[n]));
		const d_complex_t cjy = d_sub(d_rmul(+p->nz, hx[n]), d_rmul(+p->nx, hz[n]));
		const d_complex_t cjz = d_sub(d_rmul(+p->nx, hy[n]), d_rmul(+p->ny, hx[n]));

		// M = -n X E
		const d_complex_t cmx = d_sub(d_rmul(-p->ny, ez[n]), d_rmul(-p->nz, ey[n]));
		const d_complex_t cmy = d_sub(d_rmul(-p->nz, ex[n]), d_rmul(-p->nx, ez[n]));
		const d_complex_t cmz = d_sub(d_rmul(-p->nx, ey[n]), d_rmul(-p->ny, ex[n]));

		// exp(jkr * r) * dS
		const double rr = (r1[0] * p->x) + (r1[1] * p->y) + (r1[2] * p->z);
		const d_complex_t expds = d_rmul(p->ds, d_exp(kwave * rr));

		// L += M * exp(jkr * r) * dS
		plx = d_add(plx, d_mul(cmx, expds));
		ply = d_add(ply, d_mul(cmy, expds));
		plz = d_add(plz, d_mul(cmz, expds));

		// Z0 * N += (Z0 * J) * exp(jkr * r) * dS
		pnx = d_add(pnx, d_mul(cjx, expds));
		pny = d_add(pny, d_mul(cjy, expds));
		pnz = d_add(pnz, d_mul(cjz, expds));
	}

	// Z0 * N-theta, Z0 * N-phi
	const d_complex_t pnt = d_add3(d_rmul(t1[0], pnx), d_rmul(t1[1], pny), d_rmul(t1[2], pnz));
	const d_complex_t pnp = d_add3(d_rmul(p1[0], pnx), d_rmul(p1[1], pny), d_rmul(p1[2], pnz));

	// L-theta, L-phi
	const d_complex_t plt = d_add3(d_rmul(t1[0], plx), d_rmul(t1[1], ply), d_rmul(t1[2], plz));
	const d_complex_t plp = d_add3(d_rmul(p1[0], plx), d_rmul(p1[1], ply), d_rmul(p1[2], plz));

	// F-theta, F-phi
	*etheta = d_rmul(ffctr, d_add(pnt, plp));
	*ephi   = d_rmul(ffctr, d_sub(pnp, plt));
}


// alloc far field array
void alloc_farfield(void)
{
	size_t size;

	assert((Nx > 0) && (Ny > 0) && (Nz > 0) && (NFreq2 > 0));

	NSurface = 2 * ((Nx * Ny) + (Ny * Nz) + (Nz * Nx));
	//printf("%zd %d\n", NSurface, NFreq2);

	size = NSurface * sizeof(surface_t);
	Surface = (surface_t *)malloc(size);

	size = NFreq2 * sizeof(d_complex_t *);
	SurfaceEx = (d_complex_t **)malloc(size);
	SurfaceEy = (d_complex_t **)malloc(size);
	SurfaceEz = (d_complex_t **)malloc(size);
	SurfaceHx = (d_complex_t **)malloc(size);
	SurfaceHy = (d_complex_t **)malloc(size);
	SurfaceHz = (d_complex_t **)malloc(size);

	size = NSurface * sizeof(d_complex_t);
	for (int ifreq = 0; ifreq < NFreq2; ifreq++) {
		SurfaceEx[ifreq] = (d_complex_t *)malloc(size);
		SurfaceEy[ifreq] = (d_complex_t *)malloc(size);
		SurfaceEz[ifreq] = (d_complex_t *)malloc(size);
		SurfaceHx[ifreq] = (d_complex_t *)malloc(size);
		SurfaceHy[ifreq] = (d_complex_t *)malloc(size);
		SurfaceHz[ifreq] = (d_complex_t *)malloc(size);
	}
}


// setup surface E and H
void setup_farfield(void)
{
	d_complex_t cex[2][2], cey[2][2], cez[2][2];
	d_complex_t chx[2][2], chy[2][2], chz[2][2];

	// counter
	int64_t n = 0;

	// X surface
	for (int side = 0; side < 2; side++) {
		const int i = (side == 0) ? 0 : Nx;
		for (int j = 0; j < Ny; j++) {
		for (int k = 0; k < Nz; k++) {
			for (int ifreq = 0; ifreq < NFreq2; ifreq++) {
				for (int jn = 0; jn < 2; jn++) {
				for (int kn = 0; kn < 2; kn++) {
					NodeE_c(ifreq, i, j + jn, k + kn, &cex[jn][kn], &cey[jn][kn], &cez[jn][kn]);
					NodeH_c(ifreq, i, j + jn, k + kn, &chx[jn][kn], &chy[jn][kn], &chz[jn][kn]);
				}
				}
				SurfaceEx[ifreq][n] = d_complex(0, 0);
				SurfaceEy[ifreq][n] = d_rmul(0.25, d_add4(cey[0][0],
				                                          cey[0][1],
				                                          cey[1][0],
				                                          cey[1][1]));
				SurfaceEz[ifreq][n] = d_rmul(0.25, d_add4(cez[0][0],
				                                          cez[0][1],
				                                          cez[1][0],
				                                          cez[1][1]));
				SurfaceHx[ifreq][n] = d_complex(0, 0);
				SurfaceHy[ifreq][n] = d_rmul(0.25, d_add4(chy[0][0],
				                                          chy[0][1],
				                                          chy[1][0],
				                                          chy[1][1]));
				SurfaceHz[ifreq][n] = d_rmul(0.25, d_add4(chz[0][0],
				                                          chz[0][1],
				                                          chz[1][0],
				                                          chz[1][1]));
			}
			Surface[n].nx = (side == 0) ? -1 : +1;
			Surface[n].ny = 0;
			Surface[n].nz = 0;
			Surface[n].x = Xn[i];
			Surface[n].y = Yc[j];
			Surface[n].z = Zc[k];
			Surface[n].ds = (Yn[j + 1] - Yn[j]) * (Zn[k + 1] - Zn[k]);
			if (PBCx) Surface[n].ds = 0;  // X PBC -> skip X boundaries
			n++;
		}
		}
	}

	// Y surface
	for (int side = 0; side < 2; side++) {
		const int j = (side == 0) ? 0 : Ny;
		for (int k = 0; k < Nz; k++) {
		for (int i = 0; i < Nx; i++) {
			for (int ifreq = 0; ifreq < NFreq2; ifreq++) {
				for (int kn = 0; kn < 2; kn++) {
				for (int in = 0; in < 2; in++) {
					NodeE_c(ifreq, i + in, j, k + kn, &cex[kn][in], &cey[kn][in], &cez[kn][in]);
					NodeH_c(ifreq, i + in, j, k + kn, &chx[kn][in], &chy[kn][in], &chz[kn][in]);
				}
				}
				SurfaceEx[ifreq][n] = d_rmul(0.25, d_add4(cex[0][0],
				                                          cex[0][1],
				                                          cex[1][0],
				                                          cex[1][1]));
				SurfaceEy[ifreq][n] = d_complex(0, 0);
				SurfaceEz[ifreq][n] = d_rmul(0.25, d_add4(cez[0][0],
				                                          cez[0][1],
				                                          cez[1][0],
				                                          cez[1][1]));
				SurfaceHx[ifreq][n] = d_rmul(0.25, d_add4(chx[0][0],
				                                          chx[0][1],
				                                          chx[1][0],
				                                          chx[1][1]));
				SurfaceHy[ifreq][n] = d_complex(0, 0);
				SurfaceHz[ifreq][n] = d_rmul(0.25, d_add4(chz[0][0],
				                                          chz[0][1],
				                                          chz[1][0],
				                                          chz[1][1]));
			}
			Surface[n].nx = 0;
			Surface[n].ny = (side == 0) ? -1 : +1;
			Surface[n].nz = 0;
			Surface[n].x = Xc[i];
			Surface[n].y = Yn[j];
			Surface[n].z = Zc[k];
			Surface[n].ds = (Zn[k + 1] - Zn[k]) * (Xn[i + 1] - Xn[i]);
			if (PBCy) Surface[n].ds = 0;  // Y PBC -> skip Y boundaries
			n++;
		}
		}
	}

	// Z surface
	for (int side = 0; side < 2; side++) {
		const int k = (side == 0) ? 0 : Nz;
		for (int i = 0; i < Nx; i++) {
		for (int j = 0; j < Ny; j++) {
			for (int ifreq = 0; ifreq < NFreq2; ifreq++) {
				for (int in = 0; in < 2; in++) {
				for (int jn = 0; jn < 2; jn++) {
					NodeE_c(ifreq, i + in, j + jn, k, &cex[in][jn], &cey[in][jn], &cez[in][jn]);
					NodeH_c(ifreq, i + in, j + jn, k, &chx[in][jn], &chy[in][jn], &chz[in][jn]);
				}
				}
				SurfaceEx[ifreq][n] = d_rmul(0.25, d_add4(cex[0][0],
				                                          cex[0][1],
				                                          cex[1][0],
				                                          cex[1][1]));
				SurfaceEy[ifreq][n] = d_rmul(0.25, d_add4(cey[0][0],
				                                          cey[0][1],
				                                          cey[1][0],
				                                          cey[1][1]));
				SurfaceEz[ifreq][n] = d_complex(0, 0);
				SurfaceHx[ifreq][n] = d_rmul(0.25, d_add4(chx[0][0],
				                                          chx[0][1],
				                                          chx[1][0],
				                                          chx[1][1]));
				SurfaceHy[ifreq][n] = d_rmul(0.25, d_add4(chy[0][0],
				                                          chy[0][1],
				                                          chy[1][0],
				                                          chy[1][1]));
				SurfaceHz[ifreq][n] = d_complex(0, 0);
			}
			Surface[n].nx = 0;
			Surface[n].ny = 0;
			Surface[n].nz = (side == 0) ? -1 : +1;
			Surface[n].x = Xc[i];
			Surface[n].y = Yc[j];
			Surface[n].z = Zn[k];
			Surface[n].ds = (Xn[i + 1] - Xn[i]) * (Yn[j + 1] - Yn[j]);
			if (PBCz) Surface[n].ds = 0;  // Z PBC -> skip Z boundaries
			n++;
		}
		}
	}
}


// far field components
void farComponent(d_complex_t etheta, d_complex_t ephi, double e[])
{
	// abs
	e[0] = sqrt(d_norm(etheta) + d_norm(ephi));

	// theta/phi
	e[1] = d_abs(etheta);
	e[2] = d_abs(ephi);

	// major/minor
	double tmp = d_abs(d_add(d_mul(etheta, etheta), d_mul(ephi, ephi)));
	e[3] = sqrt((d_norm(etheta) + d_norm(ephi) + tmp) / 2);
	e[4] = sqrt((d_norm(etheta) + d_norm(ephi) - tmp) / 2);

	// RHCP/LHCP
	e[5] = d_abs(d_add(etheta, d_mul(d_complex(0, 1), ephi))) / sqrt(2);
	e[6] = d_abs(d_sub(etheta, d_mul(d_complex(0, 1), ephi))) / sqrt(2);
}


// far field factor
double farfactor(int ifreq)
{
	double ffctr = 0;

	const double kwave = (2 * PI * Freq2[ifreq]) / C;

	if (NFeed) {
		// feed (post only)
		double sum = 0;
		for (int ifeed = 0; ifeed < NFeed; ifeed++) {
			sum += 0.5 * Pin[MatchingLoss ? 1 : 0][(ifeed * NFreq2) + ifreq];
		}
		ffctr = kwave / sqrt(8 * PI * ETA0 * sum);
	}
	else if (IPlanewave) {
		// plane wave (solver + post)
		const double einc = 1;
		ffctr = kwave / (einc * sqrt(4 * PI));
	}

	return ffctr;
}
